core/src/client.rs

```diff
@@ -0,0 +1,231 @@
+//! Defines the `RpcClient` trait.
+use crate::{
+	Call, Error, Id, MethodCall, Output, Params,
+	Request, Response, Version,
+};
+use futures::prelude::*;
+use futures::sync::{mpsc, oneshot};
+use failure::{Fail, format_err};
+use serde_json::Value;
+use std::collections::HashMap;
+
+/// The errors returned by the client.
+#[derive(Debug, Fail)]
+pub enum RpcError {
+	/// An error returned by the server.
+	#[fail(display = "Server returned rpc error {}", _0)]
+	JsonRpcError(Error),
+	/// Failure to parse server response.
+	#[fail(display = "Failed to parse server response")]
+	ParseError,
+	/// Request timed out.
+	#[fail(display = "Request timed out")]
+	Timeout,
+	/// The server returned a response with an unknown id.
+	#[fail(display = "Server returned a response with an unknown id")]
+	UnknownId,
+	/// Not rpc specific errors.
+	#[fail(display = "{}", _0)]
+	Other(failure::Error),
+}
+
+impl From<Error> for RpcError {
+	fn from(error: Error) -> Self {
+		RpcError::JsonRpcError(error)
+	}
+}
+
+/// The future retured by the client.
+pub struct RpcFuture {
+	recv: oneshot::Receiver<Result<Value, Error>>,
+}
+
+impl RpcFuture {
+	/// Creates a new `RpcFuture`.
+	pub fn new(recv: oneshot::Receiver<Result<Value, Error>>) -> Self {
+		RpcFuture {
+			recv
+		}
+	}
+}
+
+impl Future for RpcFuture {
+	type Item = Value;
+	type Error = RpcError;
+
+	fn poll(&mut self) -> Result<Async<Self::Item>, Self::Error> {
+		// TODO should timeout
```

Comment:
This can be implemented as a wrapper, do you mind creating an issue for this and assigning a number here?